src/include/duckdb/storage/serialization/create_info.json

```diff
@@ -280,6 +280,11 @@
         "id": 206,
         "name": "cycle",
         "type": "bool"
+      },
+      {
+        "id": 207,
```

Comment:
This will probably break forward compatibility?
@Mytherin I could not manage to fix this without this change, technically we should have the information to recreate this value using `start_value`, `usage_count` and `increment`.

But CYCLE complicates this, and even without it, it won't pass this test even with this patch:
```patch
+// This should only be used for deserialization purposes, as we can guarantee that the result is within bounds in that case.
+static int64_t LastValue(CreateSequenceInfo &info) {
+       // FIXME: this is not accounting for CYCLE behavior
+       return info.start_value + (info.increment * info.usage_count);
+}
+
 SequenceData::SequenceData(CreateSequenceInfo &info)
-    : usage_count(info.usage_count), counter(info.start_value + (info.increment * info.usage_count)),
-      last_value(info.last_value), increment(info.increment), start_value(info.start_value), min_value(info.min_value),
+    : usage_count(info.usage_count), counter(LastValue(info)),
+      last_value(LastValue(info)), increment(info.increment), start_value(info.start_value), min_value(info.min_value),
       max_value(info.max_value), cycle(info.cycle) {
 }
```

Reply:
Iirc last value is supposed to be reset when the system restarts, you can double check with postgres. I would just add skipreload to the test

Reply:
I'll check postgres and add a skipreload if that's the case

Reply:
This is how it works in Postgres:

```sql
$ > psql -d postgres
psql (14.11 (Homebrew))
Type "help" for help.

postgres=# create sequence seq;
CREATE SEQUENCE
postgres=# select currval('seq');
ERROR:  currval of sequence "seq" is not yet defined in this session
postgres=# select nextval('seq');
 nextval 
---------
       1
(1 row)

postgres=# select currval('seq');
 currval 
---------
       1
(1 row)

postgres=# 
\q
$ > psql -d postgres
psql (14.11 (Homebrew))
Type "help" for help.

postgres=# select currval('seq');
ERROR:  currval of sequence "seq" is not yet defined in this session

```

Reply:
```
postgres=# select * from pg_sequences;
 schemaname | sequencename | sequenceowner | data_type | start_value | min_value |      max_value      | increment_by | cycle | cache_size | last_value 
------------+--------------+---------------+-----------+-------------+-----------+---------------------+--------------+-------+------------+------------
 public     | my_seq       | thijs         | bigint    |           5 |         3 | 9223372036854775807 |            1 | f     |          1 |          6
(1 row)

postgres=# ^D\q
➜  duckdb git:(fix_duckdb_sequences_last_value) ✗ psql postgres
psql (16.0)
Type "help" for help.

postgres=# select * from pg_sequences;
 schemaname | sequencename | sequenceowner | data_type | start_value | min_value |      max_value      | increment_by | cycle | cache_size | last_value 
------------+--------------+---------------+-----------+-------------+-----------+---------------------+--------------+-------+------------+------------
 public     | my_seq       | thijs         | bigint    |           5 |         3 | 9223372036854775807 |            1 | f     |          1 |          6
(1 row)
```

that looks persistent to me?
Am I perhaps not actually restarting since this is client-server?

Reply:
```
➜  duckdb git:(fix_duckdb_sequences_last_value) ✗ pg_ctl -D /Users/thijs/DuckDBLabs/hydra/cmake_pg_quack/../postgres-16.0/build/data stop
waiting for server to shut down.... done
server stopped
➜  duckdb git:(fix_duckdb_sequences_last_value) ✗ pg_ctl -D /Users/thijs/DuckDBLabs/hydra/cmake_pg_quack/../postgres-16.0/build/data -l /Users/thijs/DuckDBLabs/hydra/cmake_pg_quack/../postgres-16.0/build/data/postmaster.log start
waiting for server to start.... done
server started
➜  duckdb git:(fix_duckdb_sequences_last_value) ✗ psql postgres
psql (16.0)
Type "help" for help.

postgres=# select * from pg_sequences;
 schemaname | sequencename | sequenceowner | data_type | start_value | min_value |      max_value      | increment_by | cycle | cache_size | last_value 
------------+--------------+---------------+-----------+-------------+-----------+---------------------+--------------+-------+------------+------------
 public     | my_seq       | thijs         | bigint    |           5 |         3 | 9223372036854775807 |            1 | f     |          1 |          6
(1 row)
```

shut down and restarted with `pg_ctl`, still seems persisted

Reply:
Oh, strange, `currval` and `last_value` are different in Postgres. I would say we can just stick to `currval` semantics for now.

Reply:
```
postgres=# select currval('my_seq');
ERROR:  currval of sequence "my_seq" is not yet defined in this session
postgres=# insert into tbl values(default);
INSERT 0 1
postgres=# select currval('my_seq');
 currval 
---------
       7
(1 row)
```

Reply:
Or otherwise we can work on reconstructing it, but then it should work in all cases imo (including with cycle)

Reply:
Granted this behavior is postgres 16.0, not sure if 14.1 has the same behavior

Reply:
We would need to store information about the sequences inside the ClientContext if we want to replicate this behavior of "currval: sequence is not yet defined in this session" while still persisting the `usage_count`.

I'm fine adding a FIXME and slapping a skipreload on it for now

Reply:
Sounds good

Reply:
We do already persist the `usage_count`, it seems this change was made 5 years ago?
<https://github.com/duckdb/duckdb/commit/269e86bcb60083bf2c9f70921266a9fd95634509>

Anyway I'll drop it and just take the easy route for now 👍

Reply:
Actually managed to fix it without serializing, or having to reconstruct the value
We already "preserve" the last_value, by setting the `start_value` to `counter` when serializing.

Then inside the SequenceData constructor we set `usage_count` to 0 so `currvalue` throws if it hasn't been used yet in that session